display solidity framework options based on extension

technophile-04 opened the pull request:
### Description

There are major 3 scenarios: 

1. The external extension don't have any solidity  framework dir 
    Example: 
    ```shell
    yarn cli ../test-eip -e eip-712
    ```
    In this case we show all three options to user on TUI (hardhat, foundry, none)
    
2. The external extension contains only one solidity framework. 
     Example: 
     ```shell
     yarn cli ../test-eip -e technophile-04/eip-5792
     ```
     The [above extension](https://github.com/technophile-04/eip-5792/tree/main/extension/packages) only has `hardhat` so we don't give user any option on TUI and select for them. 
     
     Here there is small catch, what if in above extension user forcefully mention `foundry` using `-s` flag? 
     In this PR, preference is given to external extension directory and `-s` flag is ignored
     
3. The external extension contains both solidity framework 
     Example: 
     ```shell
     yarn cli ../test-erc -e erc-20
     ```
     In this case we show both options without the none option. Fixes #93

technophile-04 commented on src/utils/external-extensions.ts:
```diff
@@ -44,3 +45,30 @@ export const getArgumentFromExternalExtensionOption = (externalExtensionOption:
+  const ownerName = splitUrl[splitUrl.length - 2];
+  const repoName = splitUrl[splitUrl.length - 1];
+  const githubApiUrl = `https://api.github.com/repos/${ownerName}/${repoName}/contents/extension/packages${branch ? `?ref=${branch}` : ""}`;
+  const res = await fetch(githubApiUrl);
```
Since we are making unauthenticated request, github has limit on it which is 

> The primary rate limit for unauthenticated requests is 60 requests per hour.
~ [api docs](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-unauthenticated-users)

I think it should be almost good enough?

carletex replied:
We are making a a couple of requests (+ clone later) so it should be fine.

If in the future we see some issue, maybe we can just clone the repo at the beginning and read locally.

carletex commented on src/utils/external-extensions.ts:
```diff
@@ -44,3 +45,30 @@ export const getArgumentFromExternalExtensionOption = (externalExtensionOption:
+  const listOfContents = (await res.json()) as { name: string; type: "dir" | "file" }[];
+  const directories = listOfContents.filter(item => item.type === "dir").map(dir => dir.name);
+
+  const soliidtyFrameworks = Object.values(SOLIDITY_FRAMEWORKS);
```
```suggestion
  const solidityFrameworks = Object.values(SOLIDITY_FRAMEWORKS);
```

rin-st replied:
`filteredSolidityFrameworkDirs`

technophile-04 replied:
fixed at https://github.com/scaffold-eth/create-eth/pull/94/commits/3746766e99b0eacbcfb4b524b1d1f54a4bdfbb5b

carletex commented:
Works for me @rin-st 
![image](https://github.com/user-attachments/assets/be986227-1cbc-4e7a-bea5-b3dc9ad0dff6)

Sometimes I have to reset `yarn dev`. Try and see.

> Also, should it work for --dev mode? (doesn't work for me too)

I guess since `--dev` is parsing the local extension (no GitHub API request) it won't work out of the box. We could fix it by checking the folders locally (when --dev) or take the path I commented here https://github.com/scaffold-eth/create-eth/pull/94#discussion_r1705311699 (cloning the extensions at the beginning + not making any GitHub request).

Not sure if it's too important too, since it's only on --dev.

technophile-04 commented:
> I guess since --dev is parsing the local extension (no GitHub API request) it won't work out of the box. We could fix it by checking the folders locally (when --dev) or take the path I commented here https://github.com/scaffold-eth/create-eth/pull/94#discussion_r1705311699 (cloning the extensions at the beginning + not making any GitHub request).

Updated [3746766](https://github.com/scaffold-eth/create-eth/pull/94/commits/3746766e99b0eacbcfb4b524b1d1f54a4bdfbb5b) at went with kindof first, checking `typeof externalExtension === "string"` since in `--dev` mode it will be `string` and it also helps in narrowing down `externalExtension` type to object in later part of code. 

But yeah we could maybe update it later with second if people start getting rate limit error? 

> @technophile-04 What I notice is that the sorting is inverted :D Foundry is first.

Maybe it's signaling something 😂,  just fixed it at https://github.com/scaffold-eth/create-eth/pull/94/commits/df9bca6e7ff77308d565827444e188c6536cbe41

technophile-04 commented on src/utils/external-extensions.ts:
```diff
@@ -44,3 +45,30 @@ export const getArgumentFromExternalExtensionOption = (externalExtensionOption:
+  const directories = listOfContents.filter(item => item.type === "dir").map(dir => dir.name);
+
+  const soliidtyFrameworks = Object.values(SOLIDITY_FRAMEWORKS);
+  const filteredSolidityFrameworkdDirs = directories.filter(dir =>
```
Didn't get this, lol did I misspelled something again?

rin-st replied:
kdDirs -> kDirs :)

rin-st replied:
oops its fixed already

technophile-04 replied:
> kdDirs -> kDirs :)

Ohhh shit!! yup fixed it in new commit!

rin-st commented:
> Sometimes I have to reset yarn dev. Try and see.
> maybe Rinat you are missing yarn dev

Yes lol, forgot it sorry

Now it works both with `--dev` and without it, thanks!

rin-st reviewed (approved):
Lgtm!

carletex commented:
> Maybe it's signaling something 😂

lool
